internal/controller/imagepolicy_controller.go

```diff
@@ -345,6 +360,47 @@ func (r *ImagePolicyReconciler) reconcile(ctx context.Context, sp *patch.SerialP
 	return
 }
 
+func (r *ImagePolicyReconciler) updateDigest(ctx context.Context, repo *imagev1.ImageRepository, obj, oldObj *imagev1.ImagePolicy, tag string) error {
+	if obj.GetDigestReflectionPolicy() == imagev1.ReflectNever {
+		obj.Status.LatestRef.Digest = ""
+		return nil
+	}
+
+	if obj.GetDigestReflectionPolicy() == imagev1.ReflectIfNotPresent &&
+		oldObj.Status.LatestRef != nil &&
+		oldObj.Status.LatestRef.Digest != "" &&
+		obj.Status.LatestRef.Name == oldObj.Status.LatestRef.Name &&
+		obj.Status.LatestRef.Tag == oldObj.Status.LatestRef.Tag {
+		obj.Status.LatestRef.Digest = oldObj.Status.LatestRef.Digest
+		return nil
+	}
+
+	var err error
+	obj.Status.LatestRef.Digest, err = r.fetchDigest(ctx, repo, tag, obj)
+	if err != nil {
+		return fmt.Errorf("failed fetching digest of %s: %w", obj.Status.LatestRef.String(), err)
+	}
+
+	return nil
+}
+
+func (r *ImagePolicyReconciler) fetchDigest(ctx context.Context, repo *imagev1.ImageRepository, latest string, obj *imagev1.ImagePolicy) (string, error) {
+	ref := strings.Join([]string{repo.Spec.Image, latest}, ":")
+	tagRef, err := name.ParseReference(ref)
```

Comment:
This is an issue that's not introduced through this PR so I've created a separate GH issue for it: #449

Reply:
> https://github.com/fluxcd/image-reflector-controller/issues/449 doesn't seem to be a bug/issue. That's how flux controllers work.

It is a bug. Look at the `.status.latestImage` field value in the issue description. It's using the broken image reference and reports success on it. This is wrong (because obviously the tag hasn't been derived from that specific image reference) and misleading (because it doesn't tell the user that the image reference is actually wrong).